File: packages/b2c-tooling-sdk/src/clients/dual-backend-factory.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Generic factory for SCAPI/OCAPI dual backends.
 *
 * Replaces the per-domain `create*Backend()` functions (jobs, scripts,
 * users, roles) which were 100% structurally identical. Each domain now
 * supplies its constructors and config and delegates to {@link createDualBackend}.
 *
 * @module clients/dual-backend-factory
 */
import type {AuthStrategy} from '../auth/types.js';
import type {B2CInstance} from '../instance/index.js';
import {createFallbackBackend} from './scapi-fallback-backend.js';
import {resolveScapiOrOcapi, type ApiBackendPreference, type BackendBase} from './scapi-backend-utils.js';

/**
 * Common shape of every dual-backend factory's input.
 *
 * SCAPI coordinates (shortCode/tenantId) and the scope-flexible auth strategy
 * are no longer threaded in separately — they are sourced from the instance
 * via {@link B2CInstance.scapiClientConfig}. A backend is "SCAPI-capable" iff
 * that getter returns a value (shortCode + tenantId present, and a stateless
 * OAuth flow that can request the required scopes).
 *
 * `preference` is optional: when omitted it falls back to the instance's own
 * {@link B2CInstance.apiBackend} (default `'auto'`), so callers that already
 * resolved the instance from config don't have to re-plumb the flag.
 */
export interface DualBackendConfig {
  preference?: ApiBackendPreference;
  instance: B2CInstance;
}

/**
 * Configuration passed to a SCAPI backend constructor. Domains add their
 * own optional fields (e.g., `instance` for log/WebDAV access on jobs) but
 * always include shortCode + tenantId + auth.
 */
export interface ScapiBackendCtorConfig {
  shortCode: string;
  tenantId: string;
  auth: AuthStrategy;
  instance: B2CInstance;
}

/**
 * Constructors needed to build a dual-backend instance. Each domain plugs in
 * its own SCAPI/OCAPI backend classes; the factory wires them together.
 */
export interface DualBackendCtors<T extends BackendBase> {
  domainName: string;
  Scapi: new (config: ScapiBackendCtorConfig) => T;
  Ocapi: new (instance: B2CInstance) => T;
}

/**
 * Resolves the user's preference + config availability into a concrete
 * backend instance.
 *
 * - Explicit `'ocapi'` returns an OCAPI backend.
 * - Explicit `'scapi'` returns a SCAPI backend (throws if config missing).
 * - `'auto'` returns a fallback Proxy that tries SCAPI first, falls back to
 *   OCAPI on `invalid_scope`.
 *
 * @example
 * ```ts
 * export function createJobsBackend(config: JobsBackendConfig): JobsBackend {
 *   return createDualBackend(config, {
 *     domainName: 'Jobs',
 *     Scapi: ScapiJobsBackend,
 *     Ocapi: OcapiJobsBackend,
 *   });
 * }
 * ```
 */
export function createDualBackend<T extends BackendBase>(config: DualBackendConfig, ctors: DualBackendCtors<T>): T {
  const {instance} = config;
  const preference = config.preference ?? instance.apiBackend;
  const scapiClientConfig = instance.scapiClientConfig;
  const resolved = resolveScapiOrOcapi({
    preference,
    hasScapiConfig: scapiClientConfig !== undefined,
    domainName: ctors.domainName,
  });

  if (resolved === 'ocapi') {
    return new ctors.Ocapi(instance);
  }

  const scapiBackend = new ctors.Scapi({
    shortCode: scapiClientConfig!.shortCode,
    tenantId: scapiClientConfig!.tenantId,
    auth: scapiClientConfig!.auth,
    instance,
  });

  if (preference === 'scapi') {
    return scapiBackend;
  }

  // Auto mode: wrap with fallback
  const ocapiBackend = new ctors.Ocapi(instance);
  return createFallbackBackend<T>(scapiBackend, ocapiBackend, ctors.domainName.toLowerCase());
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/clients/dual_backend_factory.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Generic factory for SCAPI/OCAPI dual backends.

Mirrors ``src/clients/dual-backend-factory.ts``. Replaces the per-domain
``create_*_backend()`` functions (jobs, scripts, users, roles, ...) which were
structurally identical: each supplies its SCAPI/OCAPI backend constructors and a
config, then delegates to :func:`create_dual_backend`.

- Explicit ``"ocapi"`` returns an OCAPI backend.
- Explicit ``"scapi"`` returns a SCAPI backend (raises if SCAPI config missing).
- ``"auto"`` returns a fallback wrapper that tries SCAPI first and falls back to
  OCAPI on safe capability/auth/request rejections.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Generic, TypeVar

from b2c_tooling_sdk.clients.scapi_backend_utils import (
    ApiBackendPreference,
    BackendBase,
    ResolveBackendOptions,
    resolve_scapi_or_ocapi,
)
from b2c_tooling_sdk.clients.scapi_fallback_backend import create_fallback_backend

if TYPE_CHECKING:
    from collections.abc import Callable

    from b2c_tooling_sdk.auth.types import AuthStrategy
    from b2c_tooling_sdk.instance import B2CInstance

T = TypeVar("T", bound=BackendBase)


@dataclass
class DualBackendConfig:
    """Common input shape of every dual-backend factory.

    SCAPI coordinates and the scope-flexible auth strategy are sourced from the
    instance via :attr:`B2CInstance.scapi_client_config`. A backend is
    "SCAPI-capable" iff that getter returns a value. ``preference`` is optional:
    when ``None`` it falls back to the instance's own :attr:`B2CInstance.api_backend`
    (default ``"auto"``).
    """

    instance: B2CInstance
    preference: ApiBackendPreference | None = None


@dataclass
class ScapiBackendCtorConfig:
    """Configuration passed to a SCAPI backend constructor.

    Domains may read additional data off ``instance`` (log/WebDAV access) but
    always receive ``short_code`` + ``tenant_id`` + ``auth``.
    """

    short_code: str
    tenant_id: str
    auth: AuthStrategy
    instance: B2CInstance


@dataclass
class DualBackendCtors(Generic[T]):
    """Constructors needed to build a dual-backend instance.

    Each domain plugs in its own SCAPI/OCAPI backend factories; the generic
    factory wires them together. ``scapi`` receives a :class:`ScapiBackendCtorConfig`;
    ``ocapi`` receives the :class:`B2CInstance` directly.
    """

    domain_name: str
    scapi: Callable[[ScapiBackendCtorConfig], T]
    ocapi: Callable[[B2CInstance], T]


def create_dual_backend(config: DualBackendConfig, ctors: DualBackendCtors[T]) -> T:
    """Resolve a preference + config availability into a concrete backend instance.

    :param config: The instance + optional backend preference.
    :param ctors: Per-domain SCAPI/OCAPI constructors and domain name.
    :raises ValueError: when explicit SCAPI is requested without SCAPI config.
    """
    instance = config.instance
    preference: ApiBackendPreference = config.preference or instance.api_backend
    scapi_client_config = instance.scapi_client_config
    resolved = resolve_scapi_or_ocapi(
        ResolveBackendOptions(
            preference=preference,
            has_scapi_config=scapi_client_config is not None,
            domain_name=ctors.domain_name,
        )
    )

    if resolved == "ocapi":
        return ctors.ocapi(instance)

    assert scapi_client_config is not None  # resolve_scapi_or_ocapi guarantees this for "scapi"
    scapi_backend = ctors.scapi(
        ScapiBackendCtorConfig(
            short_code=scapi_client_config.short_code,
            tenant_id=scapi_client_config.tenant_id,
            auth=scapi_client_config.auth,
            instance=instance,
        )
    )

    if preference == "scapi":
        return scapi_backend

    # Auto mode: wrap with fallback.
    ocapi_backend = ctors.ocapi(instance)
    return create_fallback_backend(scapi_backend, ocapi_backend, ctors.domain_name.lower())


__all__ = [
    "DualBackendConfig",
    "DualBackendCtors",
    "ScapiBackendCtorConfig",
    "create_dual_backend",
]
